github/repos_contents.go

```diff
@@ -117,26 +117,30 @@ func (s *RepositoriesService) GetReadme(ctx context.Context, owner, repo string,
 // specified file. This function will work with files of any size, as opposed
 // to GetContents which is limited to 1 Mb files. It is the caller's
 // responsibility to close the ReadCloser.
-func (s *RepositoriesService) DownloadContents(ctx context.Context, owner, repo, filepath string, opts *RepositoryContentGetOptions) (io.ReadCloser, error) {
+//
+// It is possible for the download to result in a failed response. Callers
+// should check the returned http.Response status code to verify the content
+// is from a successful response.
+func (s *RepositoriesService) DownloadContents(ctx context.Context, owner, repo, filepath string, opts *RepositoryContentGetOptions) (io.ReadCloser, *http.Response, error) {
 	dir := path.Dir(filepath)
 	filename := path.Base(filepath)
 	_, dirContents, _, err := s.GetContents(ctx, owner, repo, dir, opts)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
```

Comment:
Since `s.GetContents` returns the response with its third return value, let's go ahead and return it here.

```
_, dirContents, resp, err := s.GetContents(ctx, owner, repo, dir, opts)
if err != nil {
  return nil, resp, err
```

Reply:
The response from `s.GetContents` is of type `*github.Response` rather than `*http.Response`. Do you think there is any value in returning just the `http.Response` from the `github.Response`?

We could also consider adding a 4th value to the return to allow the caller to check the rate limits from `s.GetContents` regardless of other errors?

Reply:
Ah!  Gotcha.  Hmmm... the vast majority of calls in this repo return `*github.Response`... is it a simple matter to do the same for this endpoint?  That would be ideal so that this endpoint would not be a glaring exception.
I'm thinking you can wrap the regular `*http.Response` just by calling `newResponse` (around line 438 in github.go) and returning the result.

Reply:
Sure, that looks doable. Would it make sense to return the `*github.Response` from the `s.GetContents` call with its `*http.Response` swapped out with the one from requesting the download url? I'm not very familiar with what callers are expecting from the `github.Response` but that would at least expose the rate limit information from `s.GetContents` as part of the return value.

This also makes me think it'd be useful to have a method like `RepositoriesService.GetDownloadURL(ctx context.Context, owner, repo, filepath string, opts *RepositoryContentGetOptions) (string, *github.Response, error)` that just returns the URL and leaves it up to the caller to make a request for the contents.

Reply:
Hmmm... I would return the most recent response because the point of this PR is to return the most recent response code... so I don't think we want to return anything but the response from the most recent http call, if that makes sense.

Reply:
Ok, made the change. I worry a little that this looks consistent by returning `github.Response` but may not behave consistently. The call to get the download url contents isn't an api call so we aren't able to populate any additional fields on the `github.Response` like the remaining rate limits.

Reply:
Right, but I think this is probably the best option, so let's try it out.  If it causes problems in practice, we can make another breaking API change later.  :smile:

But honestly, I like this best because the other calls that might fail could also be caused by a rate limit or some other issue that we are now exposing back to the caller.